src/workerd/jsg/value.h

```diff
@@ -851,6 +851,67 @@ class ArrayWrapper {
   }
 };
 
+// =======================================================================================
+// Sets
+
+// TypeWrapper mixin for sets.
+template <typename TypeWrapper>
+class SetWrapper {
+ public:
+  static auto constexpr MAX_STACK = 64;
+  template <typename U>
+  static constexpr const char* getName(kj::HashSet<U>*) {
+    return "Set";
+  }
+
+  template <typename U>
+  v8::Local<v8::Value> wrap(v8::Local<v8::Context> context,
+      kj::Maybe<v8::Local<v8::Object>> creator,
+      kj::HashSet<U> set) {
+    v8::Isolate* isolate = context->GetIsolate();
+    v8::EscapableHandleScope handleScope(isolate);
+
+    auto out = v8::Set::New(isolate);
+    for (const auto& item: set) {
+      out = check(out->Add(context,
+          static_cast<TypeWrapper*>(this)->wrap(context, creator, item).template As<v8::Value>()));
+    }
+
+    return handleScope.Escape(out);
+  }
+
+  template <typename U>
+  v8::Local<v8::Value> wrap(v8::Local<v8::Context> context,
+      kj::Maybe<v8::Local<v8::Object>> creator,
+      kj::HashSet<U>& set) {
+    return static_cast<TypeWrapper*>(this)->wrap(context, creator, set.asPtr());
+  }
+
+  template <typename U>
+  kj::Maybe<kj::HashSet<U>> tryUnwrap(v8::Local<v8::Context> context,
+      v8::Local<v8::Value> handle,
+      kj::HashSet<U>*,
+      kj::Maybe<v8::Local<v8::Object>> parentObject) {
+    if (!handle->IsSet()) {
+      return kj::none;
+    }
+
+    auto set = handle.As<v8::Set>();
+    auto array = set->AsArray();
+    auto length = array->Length();
+    auto builder = kj::HashSet<U>();
+    for (auto i: kj::zeroTo(length)) {
+      v8::Local<v8::Value> element = check(array->Get(context, i));
+      auto value = static_cast<TypeWrapper*>(this)->template unwrap<U>(
+          context, element, TypeErrorContext::arrayElement(i));
+      if (!builder.contains(value)) {
```

Comment:
I guess this implements "first element wins" behavior, which is _an_ option but I wonder if it's the right option. It's hard to say for sure without knowing what the future use case will be but I sort of think an exception might be better?

You can implement that by using `upsert()` and then throwing from the update callback...

Reply:
Honestly I would suggest an exception is better here. Silently choosing the first or the last and causing another entry to be dropped just smells like it will lead to subtle bugs at some point.

Reply:
I don't have strong feelings here, but I opted for "first element wins" because it seemed least surprising to me. That said, I suppose jsg conversions can fail for all sorts of reasons so it does make sense to throw in this context and it's better to be loud about it.